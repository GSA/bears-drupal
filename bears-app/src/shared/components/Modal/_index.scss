/* stylelint-disable selector-class-pattern */
.usa-modal {
  padding: 0;

  .usa-modal__content {
    padding: 64px 24px;
  }

  .usa-modal__main {
    padding: 0;
  }

  h2#modal-heading {
    font-size: 16px;
    font-weight: 600;
    line-height: 21px;
    letter-spacing: 0;
    text-align: center;
  }
}
/* stylelint-enable selector-class-pattern */
