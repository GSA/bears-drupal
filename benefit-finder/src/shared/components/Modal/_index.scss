@use '../../styles/colors/_index.scss' as color;
@use '../../styles/mixins/_index.scss' as *;
@use '../../styles/space/_index.scss' as space;

.modal-button {
  @include important-bold;

  background-color: transparent;
  border: 0;
  position: absolute;
  top: space.$space-md-plus;
  right: space.$space-md-plus;
  cursor: pointer;
}

a.nav-item-one,
a.nav-item-one {
  @include important-bold;
}

.modal-heading {
  @include important-bold;

  font-weight: 600;
  color: color.$pure-black;
  text-align: center;
  font-style: normal;
  margin: space.$space-md-plus 0;
}

.modal.usa-button-group.width-full {
  margin: 0 auto;
  text-align: center;
}
