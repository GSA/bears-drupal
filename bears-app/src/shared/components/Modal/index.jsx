import { useEffect } from 'react'
import modal from '@uswds/uswds/js/usa-modal'
import PropTypes from 'prop-types'
import Close from './assets/close.svg'
import { ObfuscatedLink, Heading } from '../index'
import './_index.scss'

/**
 * a functional component that renders a trigger, that when clicked opens a dialog
 * @component
 * @param {string} id - matches to modal control
 * @param {React.ReactNode} children - inherited children
 * @param {string} triggerLabel - passed to button for triggering modal
 * @param {string} modalHeading - heading value
 * @param {string} navItemOneLabel - passed to button for nav item in modal
 * @param {string} navItemTwoLabel - passed to button for nav item in modal
 * @return {html} returns html for seting up a usa-modal component
 */
const Modal = ({
  id,
  children,
  triggerLabel,
  modalHeading,
  navItemOneLabel,
  navItemOneFunction,
  navItemTwoLabel,
  navItemTwoFunction,
}) => {
  const handleCleanUp = () => {
    const body = document.querySelector('.usa-js-modal--active')
    if (body) {
      body.classList.remove('usa-js-modal--active')
    }
  }
  useEffect(() => {
    // initialize
    modal.on()

    // remove event listeners when the component un-mounts.
    return () => {
      modal.off()
      handleCleanUp()
    }
  }, [])

  // Have to add event listeners after components load due to the way the usa-modal works
  useEffect(() => {
    const navItemOneButton = document.querySelector('#navItemOneBtn')
    if (navItemOneButton) {
      navItemOneButton.onclick = () => navItemOneFunction()
    }

    const navItemTwoButton = document.querySelector('#navItemTwoBtn')
    if (navItemOneButton) {
      navItemTwoButton.onclick = () => navItemTwoFunction()
    }
  }, [navItemOneFunction, navItemTwoFunction])

  /**
   * a functional component that renders a link as a button for launching our dialog
   * @component
   * @param {string} id - matches to modal control
   * @param {string} triggerLabel - passed to button for triggering modal
   * @return {html} returns an obfustacted anchor element
   */
  const Trigger = ({ id, triggerLabel }) => {
    return (
      <ObfuscatedLink
        href={`#${id}`}
        aria-controls={id}
        data-open-modal
        noCarrot
      >
        {triggerLabel}
      </ObfuscatedLink>
    )
  }

  /**
   * a functional component that renders a two links as a buttons for navigating out of the dialog
   * @component
   * @param {string} navItemOneLabel - passed to button for nav item in modal
   * @param {string} navItemTwoLabel - passed to button for nav item in modal
   * @return {html} returns an obfustacted anchor element
   */
  // similar to ButtonGroup but we need links for uswds to close modal, this item is default and conditional
  const GroupNavigation = ({ navItemOneLabel, navItemTwoLabel }) => {
    return (
      <ul className="usa-button-group width-full">
        <li className="usa-button-group__item width-full">
          <ObfuscatedLink
            id="navItemOneBtn"
            className="nav-item-one width-full"
            data-close-modal
            noCarrot
          >
            {navItemOneLabel}
          </ObfuscatedLink>
        </li>
        <li className="usa-button-group__item width-full">
          <ObfuscatedLink
            id="navItemTwoBtn"
            className="nav-item-two width-full"
            data-close-modal
            noCarrot
          >
            {navItemTwoLabel}
          </ObfuscatedLink>
        </li>
      </ul>
    )
  }

  return (
    <div className="benefit-modal-group">
      <Trigger id={id} triggerLabel={triggerLabel}></Trigger>
      <div
        className="usa-modal"
        id={id}
        aria-labelledby="modal-heading"
        aria-describedby="modal-description"
      >
        <div className="usa-modal__content">
          <div className="usa-modal__main">
            <Heading headingLevel={2} id="modal-heading">
              {modalHeading}
            </Heading>
            <div className="usa-modal__footer">
              {children || (
                <GroupNavigation
                  navItemOneLabel={navItemOneLabel}
                  navItemTwoLabel={navItemTwoLabel}
                />
              )}
              {/* example: <ul className="usa-button-group">
                <li className="usa-button-group__item">
                  <button type="button" className="usa-button" data-close-modal>
                    Continue without saving
                  </button>
                </li>
                <li className="usa-button-group__item">
                  <button
                    type="button"
                    className="usa-button padding-105 text-center"
                    data-close-modal
                  >
                    Go back
                  </button>
                </li>
              </ul> */}
            </div>
          </div>
          <button
            type="button"
            className="usa-button usa-modal__close"
            aria-label="Close this window"
            data-close-modal
          >
            <img src={Close} alt="a plus icon" />
          </button>
        </div>
      </div>
    </div>
  )
}

Modal.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  triggerLabel: PropTypes.string,
  modalHeading: PropTypes.string,
  navItemOneLabel: PropTypes.string,
  navItemTwoLabel: PropTypes.string,
}

export default Modal
