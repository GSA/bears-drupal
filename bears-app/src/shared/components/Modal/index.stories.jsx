import Modal from './index.jsx'

export default {
  component: Modal,
  tags: ['autodocs'],
  args: {
    id: 'nav-modal',
    modalHeading: 'Select an option:',
    triggerLabel: 'Continue',
    navItemOneLabel: 'Verify Information',
    navItemTwoLabel: 'View Results',
  },
}

export const Primary = {}
