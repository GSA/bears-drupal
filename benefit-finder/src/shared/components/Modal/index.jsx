import { useEffect, useState, useRef } from 'react'
import NavModal from 'react-modal'
import PropTypes from 'prop-types'
import { ObfuscatedLink } from '../index'

import './_index.scss'

const customStyles = {
  overlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    width: '80%',
    borderRadius: '8px',
    borderColor: '#005ea2',
    padding: '64px 24px',
    maxWidth: '520px',
  },
}

const Close = () => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path d="M0 0h24v24H0z" fill="none" />
    <path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z" />
  </svg>
)

/**
 * a functional component that renders a trigger, that when clicked opens a dialog
 * @component
 * @param {string} id - matches to modal control
 * @param {React.ReactNode} children - inherited children
 * @param {string} triggerLabel - passed to button for triggering modal
 * @param {string} modalHeading - heading value
 * @param {string} navItemOneLabel - passed to button for nav item in modal
 * @param {string} navItemTwoLabel - passed to button for nav item in modal
 * @return {html} returns html for seting up a usa-modal component
 */
const Modal = ({
  id,
  children,
  triggerLabel,
  modalHeading,
  navItemOneLabel,
  navItemOneFunction,
  navItemTwoLabel,
  navItemTwoFunction,
  handleCheckRequriedFields,
}) => {
  // state
  const [modalIsOpen, setIsOpen] = useState(false)
  const triggerRef = useRef(null)

  // handlers
  /**
   * a function that checks for errors and then triggers the modal to open state
   * @function
   */
  const handleOpenModal = () => {
    handleCheckRequriedFields
      ? handleCheckRequriedFields() === true && setIsOpen(true)
      : setIsOpen(true)
  }

  /**
   * a function that triggers the modal to a closed state
   * @function
   */
  const handleCloseModal = triggerRef => {
    // focus the trigger if it is still in the DOM
    triggerRef && triggerRef.current.focus()
    // clear the hash
    window.location.hash = ''
    setIsOpen(false)
  }

  const handleKeyValidation = e => e.which === 32 || e.which === 13

  // effects
  useEffect(() => {
    // set our application root id here
    NavModal.setAppElement('#benefit-finder')
  }, [])

  /**
   * a functional component that renders a link as a button for launching our dialog
   * @component
   * @param {string} id - matches to modal control
   * @param {string} triggerLabel - passed to button for triggering modal
   * @return {html} returns an obfustacted anchor element
   */
  const Trigger = ({ triggerLabel, onClick, onKeyDown }) => {
    return (
      <ObfuscatedLink
        onClick={onClick}
        onKeyDown={onKeyDown}
        noCarrot
        tabIndex="0"
        triggerRef={triggerRef}
      >
        {triggerLabel}
      </ObfuscatedLink>
    )
  }

  /**
   * a functional component that renders a two links as a buttons for navigating out of the dialog
   * @component
   * @param {string} navItemOneLabel - passed to button for nav item in modal
   * @param {func} navItemOneFunction - passed to button for nav item in modal
   * @param {string} navItemOneLabel - passed to button for nav item in modal
   * @param {func} navItemTwoFunction - passed to button for nav item in modal
   * @return {html} returns an obfustacted anchor element
   */
  // similar to ButtonGroup but we need links for uswds to close modal, this item is default and conditional
  const GroupNavigation = ({
    navItemOneLabel,
    navItemOneFunction,
    navItemTwoLabel,
    navItemTwoFunction,
  }) => {
    return (
      <ul className="modal usa-button-group width-full">
        <li className="usa-button-group__item width-full" key="nav-item-one">
          <ObfuscatedLink
            id="navItemOneBtn"
            className="nav-item-one width-full"
            onClick={() => navItemOneFunction()}
            onKeyDown={e => handleKeyValidation(e) && navItemOneFunction()}
            noCarrot
            tabIndex="0"
          >
            {navItemOneLabel}
          </ObfuscatedLink>
        </li>
        <li className="usa-button-group__item width-full" key="nav-item-two">
          <ObfuscatedLink
            id="navItemTwoBtn"
            className="nav-item-two width-full"
            onClick={() => navItemTwoFunction()}
            onKeyDown={e => handleKeyValidation(e) && navItemTwoFunction()}
            noCarrot
            tabIndex="0"
          >
            {navItemTwoLabel}
          </ObfuscatedLink>
        </li>
      </ul>
    )
  }

  return (
    <div id={id} className="benefit-modal-group">
      <Trigger
        triggerLabel={triggerLabel}
        onClick={() => handleOpenModal()}
        onKeyDown={e => handleKeyValidation(e) && handleOpenModal()}
      ></Trigger>
      <NavModal
        isOpen={modalIsOpen}
        onRequestClose={handleCloseModal}
        style={customStyles}
      >
        <button
          type="button"
          className="modal-button"
          onClick={() => handleCloseModal(triggerRef)}
        >
          <Close alt="a close out icon" />
        </button>
        <div className="modal-heading">{modalHeading}</div>
        {children || (
          <GroupNavigation
            navItemOneLabel={navItemOneLabel}
            navItemOneFunction={navItemOneFunction}
            navItemTwoLabel={navItemTwoLabel}
            navItemTwoFunction={navItemTwoFunction}
          />
        )}
        {/* child example: <ul className="usa-button-group">
          <li className="usa-button-group__item">
            <button type="button" className="usa-button" data-close-modal>
              Continue without saving
            </button>
          </li>
          <li className="usa-button-group__item">
            <button
              type="button"
              className="usa-button padding-105 text-center"
              data-close-modal
            >
              Go back
            </button>
          </li>
        </ul> */}
      </NavModal>
    </div>
  )
}

Modal.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  triggerLabel: PropTypes.string,
  modalHeading: PropTypes.string,
  navItemOneLabel: PropTypes.string,
  navItemOneFunction: PropTypes.func,
  navItemTwoLabel: PropTypes.string,
  navItemTwoFunction: PropTypes.func,
}

export default Modal
