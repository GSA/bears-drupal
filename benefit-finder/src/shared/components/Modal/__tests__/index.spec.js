import renderer from 'react-test-renderer'
import Modal from '../index.jsx'

describe('Modal', () => {
  it('renders a match to the previous snapshot', () => {
    const component = renderer.create(<Modal />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
