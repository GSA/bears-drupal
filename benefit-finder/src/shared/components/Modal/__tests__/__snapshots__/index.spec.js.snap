// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders a match to the previous snapshot 1`] = `
<div
  className="benefit-modal-group"
>
  <a
    className=""
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex="0"
  />
</div>
`;
